control-plane/pkg/contract/subscriptionsapi.go

```diff
@@ -0,0 +1,196 @@
+	}
+	for _, f := range filters {
+		all.Filters = append(all.Filters, FromSubscriptionFilter(f))
+	}
+	return &DialectedFilter{
+		Filter: &DialectedFilter_All{
+			All: &all,
+		},
+	}
+}
+
+// NewAnyFilter converts the SubscriptionsAPIFilter into the any dialect of the
+// DialectedFilter as defined in CloudEvents Subscriptions API.
+//
+// All filter evaluates to true when all nested filters evaluate to true.
+//
+// See "CNCF CloudEvents Subscriptions API" > "3.2.4.1 Filters Dialects"
+// https://github.com/cloudevents/spec/blob/main/subscriptions/spec.md#any-filter-dialect
+func NewAnyFilter(filters []v1.SubscriptionsAPIFilter) *DialectedFilter {
+	any := Any{
+		Filters: []*DialectedFilter{},
+	}
+	for _, f := range filters {
+		any.Filters = append(any.Filters, FromSubscriptionFilter(f))
+	}
+	return &DialectedFilter{
+		Filter: &DialectedFilter_Any{
+			Any: &any,
+		},
+	}
+}
+
+// NewNotFilter converts the SubscriptionsAPIFilter into the not dialect of the
+// DialectedFilter as defined in CloudEvents Subscriptions API.
+//
+// Not filter evaluates to true when the nested filter evaluates to false.
+//
+// See "CNCF CloudEvents Subscriptions API" > "3.2.4.1 Filters Dialects"
+// https://github.com/cloudevents/spec/blob/main/subscriptions/spec.md#not-filter-dialect
+func NewNotFilter(f v1.SubscriptionsAPIFilter) *DialectedFilter {
+	return &DialectedFilter{
+		Filter: &DialectedFilter_Not{
+			Not: &Not{
+				Filter: FromSubscriptionFilter(f),
+			},
+		},
+	}
+}
+
+// NewCESQLFilter converts the SubscriptionsAPIFilter into the sql dialect of the
+// DialectedFilter as defined in CloudEvents Subscriptions API.
+//
+// CESOL filter is a Cloud Events SQL Expression
+//
+// See "CNCF CloudEvents Subscriptions API" > "3.2.4.1 Filters Dialects"
+// https://github.com/cloudevents/spec/blob/main/subscriptions/sql.md#any-filter-dialect
+//
+// See "CNCF CloudEvents SQL Expression Language"
+// https://github.com/cloudevents/spec/blob/main/cesql/spec.md
+func NewCESQLFilter(f v1.SubscriptionsAPIFilter) *DialectedFilter {
+	return &DialectedFilter{
+		Filter: &DialectedFilter_Cesql{
+			Cesql: &CESQL{
+				Expression: f.SQL,
+			},
+		},
+	}
+}
+
+// FromSubscriptionFilter converts a SubscriptionsAPIFilter to the corresponding
+// DialectedFilter based on the property set in the SubscriptionsAPIFilter.
+//
+// E.g a SubscriptionsAPIFilter with the All field set will be converted to a DialectedFilter
+// which Filter is of type DialectedFilter_All.
+func FromSubscriptionFilter(f v1.SubscriptionsAPIFilter) *DialectedFilter {
+	switch {
+	case len(f.All) > 0:
+		return NewAllFilter(f.All)
+	case len(f.Any) > 0:
+		return NewAnyFilter(f.Any)
+	case f.Not != nil:
+		return NewNotFilter(*f.Not)
+	case f.Exact != nil:
+		return NewExactFilter(f)
+	case f.Prefix != nil:
+		return NewPrefixFilter(f)
+	case f.Suffix != nil:
+		return NewSuffixFilter(f)
+	case f.SQL != "":
+		return NewCESQLFilter(f)
+	}
+	return &DialectedFilter{}
+}
```

Comment:
Can we decouple the one filter constraint from the contract?
I remember making this comment in another PR :thinking:

Reply:
In addition, the CE subscription spec doesn't have this constraint, so I would really avoid designing a contract in this way

Reply:
> Can we decouple the one filter constraint from the contract? I remember making this comment in another PR thinking

Yes I'm working on it. 

> In addition, the CE subscription spec doesn't have this constraint
Yeah while it's not in the spec itself, it's in the reference openapi file and I remember that it was brought up in one of the PRs introducing the spec. But I agree that we shouldn't have it tightly coupled here.
https://github.com/cloudevents/spec/blob/main/subscriptions/subscriptions-openapi.yaml#L175-L183

Reply:
Fixed